src/Simulation/Native/src/simulator/dbw_test.cpp

```diff
@@ -0,0 +1,231 @@
+    std::vector<std::int32_t> empty;
+    string line;
+    ifstream file(fName);
+    if (!file.is_open()) throw(std::invalid_argument("Can't open input file"));
+
+    int phase = 0;
+    if (doClusters) phase = 2;
+
+    regex reOrig("^=== Original:.*[\r]?");
+    regex reGate("^\\s*(\\d+):\\s+(.+)\\[(.*)\\].*[\r]?");
+    regex reClusters("^=== Clusters.*[\r]?");
+    regex reCluster("^==== cluster\\[\\s*(\\d+)\\]:.*[\r]?");
+    smatch sm;
+
+    while (getline(file, line)) {
+        if (phase == 99) break;
+        switch (phase) {
+        case 0:
+            if (regex_match(line, sm, reOrig)) phase = 1;
+            break;
+        case 1:
+            if (regex_match(line, sm, reGate)) {
+                auto qs = splitNums(sm[3]);
+                rslt.push_back(qs);
+            }
+            else phase = 99;
+            break;
+        case 2:
+            if (regex_match(line, reClusters)) 
+                phase = 3;
+            break;
+        case 4:
+            if (regex_match(line, sm, reGate)) {
+                auto qs = splitNums(sm[3]);
+                rslt.push_back(qs);
+                break;
+            }
+            else {
+                phase = 3;
+                [[fallthrough]];
+            }
+        case 3:
+            if (regex_match(line, sm, reCluster)) {
+                rslt.push_back(empty);
+                phase = 4;
+            }
+            break;
+        }
+    }
+    file.close();
+    return rslt;
+}
+
+void mySprintf(char* buf, int bufSiz, const char* fmt, ...) {
+    va_list args;
+#ifdef _MSC_VER
+    __crt_va_start(args, fmt);
+    vsprintf_s(buf, bufSiz, fmt, args);
+    __crt_va_end(args);
+#else
+    va_start(args,fmt);
+    vsprintf(buf, fmt, args);
+    va_end(args);
+#endif
+    //perror(buf);
+}
+
+int numQs(vector<vector<int32_t>> prb) {
+    int mx = -1;
+    for (auto i : prb)
+        for (auto j : i)
+            if (j > mx) mx = j;
+    return (mx + 1);
+}
+
+int main()
+{
+    int                     nQs;
+    vector<vector<int32_t>> prb;
+    char                    fName[30];
+
+    // Perform a small number of loops on the 4x4 advantage circuit.
+    int sizR = 4;
+    int sizC = 4;
+    int loops = 10;
+    mySprintf(fName, sizeof(fName), "advantage_%d%d_4.log", sizR, sizC);
+
+    prb         = loadTest(fName, false);
+    nQs         = numQs(prb);
+    int gateCnt = (int)prb.size();
+
+    printf("==== Starting %s (%d gates)\n", fName, gateCnt);
+
+    auto sim_id = init();
+    for (int q = 0; q < nQs; q++) allocateQubit(sim_id, q);
+
+    std::chrono::system_clock::time_point start = std::chrono::system_clock::now();
+    int itvl = loops / 10;
+    for (int i = 0; i < loops; i++) {
+        for (int j = 0; j < prb.size(); j++) {
+            auto qs = prb[j];
+            uint32_t cs[2];
+            switch (qs.size()) {
+            case 0: // Need to force a flush (end of cluster)
+                //Flush(sim_id);
```

Comment:
@dbwz8 Why is this flush commented out? I noticed in the debug version of this code (now removed) the flush was not commented out, and the comment above this seems to indicate that the force flush is needed. Which is correct?

Reply:
This is for testing without a scheduler (when clusters are pre-computed and loaded from a file). The real scheduler does the flush at the correct time (at the end of each cluster). This Flush is only necessary during DBWDBG (when we might be testing without the real scheduler). In fact, the Flush() call is only defined if DBWDBG is defined, so it had to be commented out. It can be removed in the "real" code.

Reply:
Thanks for the explanation! I've removed the commented out lines.